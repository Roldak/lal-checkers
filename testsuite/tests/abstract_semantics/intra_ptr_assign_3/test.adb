procedure Test is
   type Point is record
      x : Integer;
      y : Integer;
   end record;

   b : Boolean;
   x : Point := (3, 1);
   y : Point := (4, 1);

   ptr : access Integer;
begin
   if b then
      ptr := x.x'Access;
   else
      ptr := y.x'Access;
   end if;

   ptr.all := 5;
end Ex1;
