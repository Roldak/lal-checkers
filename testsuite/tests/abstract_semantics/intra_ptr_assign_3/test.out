{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assume0",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{True}",
          "ptr": "{Get_0(0x00000002)}"
        }
      }
    ],
    "assign5": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign5",
          "assume1",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False}",
          "ptr": "{Get_0(0x00000004)}"
        }
      }
    ],
    "assign6": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assume0",
          "assume1",
          "read0",
          "read1",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 5], [1, 1])', 4: '([4, 5], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{Get_0(0x00000002), Get_0(0x00000004)}"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assume0",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assume1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assume1",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "read1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "split_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "read0",
          "read1",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{Get_0(0x00000002), Get_0(0x00000004)}"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ]
  },
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assign4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assume0",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{True}",
          "ptr": "{Get_0(0x00000002)}"
        }
      }
    ],
    "assign5": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign5",
          "assume1",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False}",
          "ptr": "{Get_0(0x00000004)}"
        }
      }
    ],
    "assign6": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign6",
          "assume0",
          "read0",
          "read1",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([5, 5], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{True}",
          "ptr": "{Get_0(0x00000002)}"
        }
      },
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign5",
          "assign6",
          "assume1",
          "read0",
          "read1",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([5, 5], [1, 1])'}, 0)",
          "b": "{False}",
          "ptr": "{Get_0(0x00000004)}"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assume0",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "assume1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assume1",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "read1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "read0",
          "read1",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ],
    "split_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assume0",
          "read0",
          "read1",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{True}",
          "ptr": "{Get_0(0x00000002)}"
        }
      },
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign5",
          "assume1",
          "read0",
          "read1",
          "split_join0",
          "start0"
        ],
        "values": {
          "$stack": "({2: '([3, 3], [1, 1])', 4: '([4, 4], [1, 1])'}, 0)",
          "b": "{False}",
          "ptr": "{Get_0(0x00000004)}"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "$stack": "({}, 0)",
          "b": "{False, True}",
          "ptr": "{[all-path]}"
        }
      }
    ]
  }
}